napari/plugins/_npe2.py

```diff
@@ -282,11 +294,115 @@ def _on_plugin_enablement_change(enabled: Set[str], disabled: Set[str]):
+    """
+    from .._app import app
+
+    actions, submenus = _npe2_manifest_to_actions(manifest)
+    if actions:
+        disposable = app.register_actions(actions)
+        pm.get_context(manifest.name).register_disposable(disposable)
+    if submenus:
+        disposable = app.menus.append_menu_items(submenus)
+        pm.get_context(manifest.name).register_disposable(disposable)
+
+
+def _is_menu_contributable(menu_id) -> bool:
+    """Return True if the given menu_id is a menu that plugins can contribute to."""
+    # here is where we can check whether the menu id is "contributable"
+    # or not.  i.e. if they're trying to add to the File menu, we skip it here.
+    from .._app.constants import MenuId
+
+    if menu_id.startswith("napari/"):
+        # emit a warning?
+        return menu_id in {MenuId.LAYERLIST_CONTEXT.value}
+    return True
+
+
+def _npe2_manifest_to_actions(
+    mf: PluginManifest,
+) -> Tuple[List[Action], List[Tuple[str, SubmenuItem]]]:
+    """Gather actions and submenus from a npe2 manifest, export app_model types."""
+    from app_model.types import Action, MenuRule
+
+    cmds: DefaultDict[str, List[MenuRule]] = DefaultDict(list)
+    submenus: List[Tuple[str, SubmenuItem]] = []
+    for menu_id, items in mf.contributions.menus.items():
+        if not _is_menu_contributable(menu_id):
+            continue  # pragma: no cover
+        for item in items:
+            if isinstance(item, contributions.MenuCommand):
+                rule = MenuRule(id=menu_id, **_when_group_order(item))
+                cmds[item.command].append(rule)
+            else:
+                submenus.append((menu_id, _npe2_submenu_to_app_model(item)))
+
+    actions = [
+        Action(
+            id=cmd.id,
+            title=cmd.title,
+            category=cmd.category,
+            tooltip=cmd.short_title or cmd.title,
+            icon=cmd.icon,
+            enablement=cmd.enablement,
+            callback=cmd.python_name,
+            menus=cmds.get(cmd.id),
+            keybindings=[],
+        )
+        for cmd in mf.contributions.commands or ()
+    ]
+    return actions, submenus
+
+
+def _when_group_order(
+    menu_item: contributions.MenuItem,
+) -> dict[str, Union[str, float, None]]:
+    """Extract when/group/order from an npe2 Submenu or MenuCommand."""
+    group, _, _order = (menu_item.group or '').partition("@")
+    try:
+        order: Optional[float] = float(_order)
+    except ValueError:
+        order = None
+    return {'when': menu_item.when, 'group': group or None, 'order': order}
+
+
+def _npe2_submenu_to_app_model(subm: contributions.Submenu) -> SubmenuItem:
+    """Convert a npe2 submenu contribution to an app_model SubmenuItem."""
+    contrib = pm.get_submenu(subm.submenu)
+    return SubmenuItem(
+        submenu=contrib.id,
+        title=contrib.label,
+        icon=contrib.icon,
+        **_when_group_order(subm),
+        # enablement= ??  npe2 doesn't have this, but app_model does
```

Comment:
foreshadowing I hope, plugins which can make things available in specific contexts feels pretty huge! I guess enablement for plugin contributions should automatically include 'when this plugin is active'? 🙂

Reply:
>  I hope, plugins which can make things available in specific contexts feels pretty huge!

yes.  Actually that's already very possible with the [`command.enablement` field](https://github.com/napari/npe2/blob/85c7d23c4a7ed1011a044456b21fa4a68e15a10c/npe2/manifest/contributions/_commands.py#L82), which determines whether some callable is appropriate given the context (it was only just recently uncommented in npe2, so no existing plugins have populated it).

This particular commented out field here only pertains to whether a _submenu_ contribution can be grayed out or not depending on the context.  As far as I can tell, vscode didn't have this anywhere: submenus are _always_ active, even if all of their items are not.  but i think we want to be able to gray out a whole submenu